Add documentation for MetadataElement

bernt-matthias opened the pull request:
In https://github.com/galaxyproject/tools-iuc/issues/4367 we realised that documentation is missing for metadata elements of datasets. 

Places in the code corresponding to statements in the new docs:

`default`

- `default` is used if not set (by `set_meta` or the user): probably here:  https://github.com/galaxyproject/galaxy/blob/6df8d1ae5b54bbb0961576622066a3d2b95f6728/lib/galaxy/model/metadata.py#L73

`no_value`

-  in cheetah https://github.com/galaxyproject/galaxy/blob/6df8d1ae5b54bbb0961576622066a3d2b95f6728/lib/galaxy/tools/wrappers.py#L283
- metadata validator https://github.com/bernt-matthias/galaxy/blob/347e9fed2ea77576aebe5256151e8c493daacdd1/lib/galaxy/datatypes/data.py#L246 .. but this has been introduced recently https://github.com/galaxyproject/galaxy/pull/13139
- filter https://github.com/galaxyproject/galaxy/blob/6df8d1ae5b54bbb0961576622066a3d2b95f6728/lib/galaxy/tools/parameters/dynamic_options.py#L207

## How to test the changes?
(Select all options that apply)
- [ ] I've included appropriate [automated tests](https://docs.galaxyproject.org/en/latest/dev/writing_tests.html).
- [ ] This is a refactoring of components with existing test coverage.
- [ ] Instructions for manual testing are as follows:
  1. [add testing steps and prerequisites here if you didn't write automated tests covering all your changes]

## License
- [x] I agree to license these contributions under [Galaxy's current license](https://github.com/galaxyproject/galaxy/blob/dev/LICENSE.txt).
- [x] I agree to allow the Galaxy committers to license these and all my past contributions to the core galaxy codebase under the [MIT license](https://opensource.org/licenses/MIT). If this condition is an issue, uncheck and just let us know why with an e-mail to galaxy-committers@lists.galaxyproject.org.

bernt-matthias committed:
add documentation for MetadatElement

mvdbeek commented on doc/source/dev/data_types.md:
```diff
@@ -246,14 +246,34 @@ entries inside your class like this:
+Here we have a `MetadataElement`, accessible in methods with a dataset parameter
+from `dataset.metadata.number_of_sequences`. 
+
+- `default` (default is `None`): the default value that is set if neither
```
Is that right ? Where's that happening ?

bernt-matthias replied:
I guess here https://github.com/galaxyproject/galaxy/blob/6df8d1ae5b54bbb0961576622066a3d2b95f6728/lib/galaxy/model/metadata.py#L134 see also docstring here https://github.com/galaxyproject/galaxy/blob/6df8d1ae5b54bbb0961576622066a3d2b95f6728/lib/galaxy/model/metadata.py#L73

mvdbeek commented on doc/source/dev/data_types.md:
```diff
@@ -246,14 +246,34 @@ entries inside your class like this:
+- `default` (default is `None`): the default value that is set if neither
+  `set_meta` (see below) nor the user (if allowed, see `readonly`) has set a value
+- `no_value` (defaults to `None`): a value that indicates that the metadata has
+  not been set. In most cases `no_value` should be an invalid value, in particulat
```
Is that true ? no_value seems to be used as the value if the metadata value is None.

bernt-matthias replied:
Not clear to which part your question relates: 

>  a value that indicates that the metadata has not been set

This is what I understood from the initial commit message https://github.com/galaxyproject/galaxy/commit/584999c5f0da243c2a936406e7034b7e2b39e02b

> In most cases `no_value` should be an invalid value

Otherwise, i.e. if it is a valid/possible value in the context of the datatype, one can not find out if the value is set.

mvdbeek commented on doc/source/dev/data_types.md:
```diff
@@ -246,14 +246,34 @@ entries inside your class like this:
+  `set_meta` (see below) nor the user (if allowed, see `readonly`) has set a value
+- `no_value` (defaults to `None`): a value that indicates that the metadata has
+  not been set. In most cases `no_value` should be an invalid value, in particulat
+  one that can not be set by `set_meta`. In general `no_value` should be
```
I don't think that's true for optional elements ?

bernt-matthias replied:
> In most cases `no_value` should be an invalid value, in particulat one that can not be set by `set_meta`

Also optional metadata can be set by `set_meta` (maybe its just not possible for all datasets) or the user. If we allow valid (in the context of the datatype) values then we do not know if the metadata is set if it has this value.

bernt-matthias replied:
> In general `no_value` should be   different from `default`

Same argumentation.

bernt-matthias committed:
add another use of no_value

and more parameters that are actually used, but I have no idea (yet)
what to say about them

bernt-matthias committed:
add possisble param

mvdbeek commented on doc/source/dev/data_types.md:
```diff
@@ -246,14 +246,42 @@ entries inside your class like this:
+- `no_value` (defaults to `None`): a value that indicates that the metadata has
+  not been set. In most cases `no_value` should be an invalid value, in particular
+  one that can not be set by `set_meta`. In general `no_value` should be
+  different from `default`. Having a `no_value` different from `None` can be of
```
Why is this a good example ?

mvdbeek commented on doc/source/dev/data_types.md:
```diff
@@ -246,14 +246,42 @@ entries inside your class like this:
+     set (note that such a validator is added implicitly for all dataset
+     parameters),
+  2. `no_value` is used in the Cheetah code for metadata whose value is `None`
+  3. For metadata filters (applied to dynamic options), where unset
```
Why do we do that ?

mvdbeek commented on doc/source/dev/data_types.md:
```diff
@@ -246,14 +246,42 @@ entries inside your class like this:
+  1. for the metadata validator that verifies that all mandatory metadata data is
+     set (note that such a validator is added implicitly for all dataset
+     parameters),
+  2. `no_value` is used in the Cheetah code for metadata whose value is `None`
```
I am not sure this is really what that code is meant to do. The metadata parameters `to_string` uses no_value also for `[]`
